### PR TITLE
Implement ksp

### PR DESCRIPTION
**Pull request by BraisGabin**

### CODE REVIEW

File: lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt [hunk 1]
```diff
@@ -66,23 +79,36 @@ internal fun BindingGraph.ComponentNode.getDependenciesCodePosition(elements: El
...
         .fold(
             { element ->
                 element.enclosedElements
                     .filter {
-                        it.isAnnotatedWith(Component.Factory::class) || it.isAnnotatedWith(Subcomponent.Factory::class) ||
-                            it.isAnnotatedWith(Component.Builder::class) || it.isAnnotatedWith(Subcomponent.Builder::class)
+                        it.isAnnotatedWith(Component.Factory::class) ||
+                            it.isAnnotatedWith(Subcomponent.Factory::class) ||
+                            it.isAnnotatedWith(Component.Builder::class) ||
+                            it.isAnnotatedWith(Subcomponent.Builder::class)
```
Comment (BraisGabin) [inline lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt:97]: Could we make this easier? a list and `any`?

File: lightsaber/src/main/java/schwarz/it/lightsaber/utils/ComponentNode.kt [hunk 2]
```diff
@@ -17,14 +17,26 @@ internal fun BindingGraph.ComponentNode.getDeclaredModules(
...
-            .map { Module(types.asElement(it) as TypeElement) }
+        componentPath().currentComponent()
+            .fold(
+                { element ->
+                    element.getAnnotation(Subcomponent::class.java)
+                        .getTypesMirrorsFromClass { modules }
+                        .map { Module(types.asElement(it) as TypeElement) }
+                },
+                { it.getDeclaredModules(dagger.Subcomponent::class, "modules") },
+            )
```
Comment (BraisGabin) [inline lightsaber/src/main/java/schwarz/it/lightsaber/utils/ComponentNode.kt:31]: Could we remove this if and merge the code for Component and subcomponent?

File: lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt [hunk 3]
```diff
@@ -66,23 +79,36 @@ internal fun BindingGraph.ComponentNode.getDependenciesCodePosition(elements: El
...
 
+@OptIn(KspExperimental::class)
 internal fun BindingGraph.ComponentNode.getComponentFactoriesAndBuilders(): List<FactoryOrBuilder> {
     return componentPath()
         .currentComponent()
         .fold(
             { element ->
                 element.enclosedElements
                     .filter {
-                        it.isAnnotatedWith(Component.Factory::class) || it.isAnnotatedWith(Subcomponent.Factory::class) ||
```
Comment (javils) [inline lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt:80]: Extract this duplicated code into a list of annotations

**Review by javils — CHANGES_REQUESTED** [review]

File: lightsaber/src/main/java/schwarz/it/lightsaber/domain/Dependency.kt [hunk 4]
```diff
@@ -9,7 +10,11 @@ interface Dependency {
 
     companion object {
         operator fun invoke(element: DaggerElement): Dependency {
-            return element.fold(::DependencyJavac, { TODO("ksp is not supported yet") })
+            return element.fold(::DependencyJavac, { DependencyKsp(it as KSDeclaration) })
```
[the remark this answers is not in the document]
Comment (raul19) [reply lightsaber/src/main/java/schwarz/it/lightsaber/domain/Dependency.kt:13]: We can not change the type.

File: lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt (hunk 3 above)
Comment (raul19) [reply lightsaber/src/main/java/schwarz/it/lightsaber/utils/BindingGraph.kt:80]: Done.

**Review by raul19 — APPROVED** [review]

**Review by javils — APPROVED** [review]

**Pull request merged**